### PR TITLE
Add a ProwJob queue functionality to plank

### PR DESCRIPTION
**Pull request by jupblb**
Allow jobs to be aggregated in a queue that secures concurrent execution. This can be especially useful when coupled with boskos pools. Fixes https://github.com/kubernetes/test-infra/issues/21996.

### CODE REVIEW

**Comment by jupblb** [comment]
For now this is just a proposition based on https://github.com/kubernetes/test-infra/issues/21996#issuecomment-828615019. It would use labels defined in a `ProwJob` to establish the name and size of desired queue.

The target is to use this together with some Boskos pools and the number of the related GCP projects would determine the queue max concurrency (each boskos pool would be matched with a different queue name).

The worst thing about this idea is that the limit would be defined by each `ProwJob` (separately) instead of having it set globally. This should be changed but at this moment I do not have a very good idea what is the cleanest solution here.

/cc @mborsz @chaodaiG

**Comment by alvaroaleman** [comment]
Sorry, I never provided feedback for this. I like the approach but I am also the one who suggested it, so I might be biased. Using prow config to store the  queuename-> concurrency mapping makes sense.

It would maybe be nice to actually have a spec field for this, so it is easier to discover (and default to an empty string which will be a queue with indefinite concurrency). From an implementation POV the label is better, as we can filter by label but we can simply copy the spec field to the label.

Would like to hear @chaodaiG and @cjwagner thoughts

**Comment by cjwagner** [comment]
> It would maybe be nice to actually have a spec field for this, so it is easier to discover (and default to an empty string which will be a queue with indefinite concurrency). From an implementation POV the label is better, as we can filter by label but we can simply copy the spec field to the label.

I prefer the idea of using a spec field since it can be more easily documented/discovered and validated. ~I would expect an empty string default to retain the existing behavior of using the job name as the queue key rather than allowing unlimited concurrency (that would be backwards incompatible).~ Actually it looks like this is a completely separate concurrency mechanism from the job name based max concurrency setting, if that is the case please clarify this in the documentation for all 3 fields.

Rather than copying the queue key to a label and having two sources of truth I'd prefer if we just added a cache index for the field. Jobs have a lot of labels now and we only need this index in prow-controller-manager.

**Comment by jupblb** [comment]
I agree with this feedback. I made another commit with suggested fixes. `ProwJobSpec.JobQueueName` is now used instead of a custom label. Please let me know if this is about right. If so then I'll add some tests, squash the commits and make this ready for review. Thanks!

File: prow/apis/prowjobs/v1/types.go
Comment (cjwagner) [inline prow/apis/prowjobs/v1/types.go:206]: There is currently no way to configure this field for a job because it does not exist in the job configuration, only the ProwJob resource. We need to add this field to configuration (specifically `JobBase`) and make sure it is copied to the ProwJob resource when [creating the ProwJobSpec](https://github.com/kubernetes/test-infra/blob/61275e2d9c3f4acad2cd5383f1ad174e9078108d/prow/pjutil/pjutil.go#L192). You can grep for `MaxConcurrency` to see how this should be done.

File: prow/plank/reconciler.go
Comment (cjwagner) [inline prow/plank/reconciler.go:813]: We shouldn't allow/expect the `!queueDefined` case. Please add job config validation logic to the `config` package to ensure that all job queue names specified on jobs are associated with an entry in the `JobQueueConcurrencies` map. Including the validation logic there will catch such misconfigurations at presubmit time rather than runtime.

File: prow/plank/reconciler.go
Comment (cjwagner) [inline prow/plank/reconciler.go:818]: `optPendingProwJobs()` seems overly restrictive. That will omit triggered ProwJobs and cause us to undercount in `countPendingOrOlderMatchingPJs` (there will be no older triggered PJs).

File: prow/plank/reconciler.go
Comment (cjwagner) [inline prow/plank/reconciler.go:98]: Please augment the existing `prowJobIndexer` rather than defining another.

File: prow/plank/reconciler.go
Comment (cjwagner) [inline prow/plank/reconciler.go:929]: It is dangerous to use the raw user supplied value as the index key here. There would likely be problems if the user supplied value matched an index key we define such as `all`, `pending`, or `pending-triggered-named-foo`. Please prepend a prefix like `job-queue-name` to prevent this.

File: prow/plank/reconciler.go
Comment (alvaroaleman) [reply prow/plank/reconciler.go:98]: @cjwagner this adds a new index, using the existing index is possible but requires some kind of prefix to avoid collisions (as you mentioned inside the indexer func) - What drawbacks do you see with adding a new index?

Comment (cjwagner) [reply prow/plank/reconciler.go:98]: I didn't realize that a separate index name was being used as an alternative to using a value prefix, that makes more sense. I was also thinking of using a prefix for consistency with the existing indexing logic and because I wasn't sure about the performance cost of additional indexers. Both these concerns are negligible though.

File: prow/apis/prowjobs/v1/types.go
Comment (jupblb) [reply prow/apis/prowjobs/v1/types.go]: Fixed in https://github.com/kubernetes/test-infra/pull/25674/commits/82d5fddb4ab98d9e00e593d47e555546a5a04216.

File: prow/apis/prowjobs/v1/types.go
Comment (jupblb) [reply prow/apis/prowjobs/v1/types.go:209]: Fixed in https://github.com/kubernetes/test-infra/pull/25674/commits/6cf071713ef2ccd31473141cb753e2528d26e202.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: Fixed in https://github.com/kubernetes/test-infra/pull/25674/commits/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd. Added a `pendingTriggeredIndexKeyByJobQueueName` selector.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: I believe that reusing the existing indexer is indeed a cleaner approach so I changed the code to do just that. Fixed in https://github.com/kubernetes/test-infra/pull/25674/commits/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: Since this was using another indexer I don't think this would be a problem? But either way this is no longer the case as of https://github.com/kubernetes/test-infra/pull/25674/commits/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd (this function is removed).

**Comment by jupblb** [comment]
Thank you for the feedback. I applied the requested changes. To make the review process easier I kept them in separate commits. Again, once given a green light I'll add some tests, squash the commits and make this ready for final review.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go:810]: If I understand correctly, the validation happens when a `JobConfig` is converted to a `ProwJob`. When a user does that locally, they can have the names already added to the prow configuration and the validation would pass. But that doesn't change the fact that if the server has outdated plank configuration, it would fail to find queue concurrency for given `queueName`. So we should anticipate for such a scenario either way, right?

Comment (jupblb) [reply prow/plank/reconciler.go:810]: Another thing is that adding this validation logic to https://github.com/kubernetes/test-infra/blob/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd/prow/config/config.go#L1905-L1943 might be a bit... "dirty". We would need to pass the `c.ProwConfig.Plank.JobQueueConcurrencies` down to `validatePresubmits`, `validatePostsubmits` and `validatePeriodics`.

File: prow/config/jobs.go
Comment (cjwagner) [inline prow/config/jobs.go:131]: `s/0/omission/`

File: prow/plank/reconciler.go
Comment (cjwagner) [reply prow/plank/reconciler.go:810]: > If I understand correctly, the validation happens when a `JobConfig` is converted to a `ProwJob`

No, config validation happens when the config is loaded. We load the config as part of `checkconfig`'s validation at presubmit time to prevent submission of invalid config.

> When a user does that locally, they can have the names already added to the prow configuration and the validation would pass. But that doesn't change the fact that if the server has outdated plank configuration, it would fail to find queue concurrency for given queueName.

Job configuration and the core config are both updated immediately after the config change PR is merged. Plank should never have configuration that is outdated with respect to the job configuration by more than a few seconds and that can only happen after a job is added or queue name is changed.

> So we should anticipate for such a scenario either way, right?

We do need to gracefully handle this case. We don't want plank to explode in the rare event of a job being triggered and transitioning to pending immediately after the job is added and after the job config has been updated, but before the core config has been updated. We should log a debug level message and probably abort the job.

> Another thing is that adding this validation logic to [...] might be a bit... "dirty".

Yes, we'll have to do something like you suggested since the `JobQueueConcurrencies` is not available in `validateJobBase` currently, but the dirtiness is definitely worth the UX improvement of having the validation to catch this at presubmit time. Either of the approaches you mentioned seems appropriate for now, we can refactor later if we find more fields need to be passed down.

File: prow/plank/reconciler.go
Comment (cjwagner) [inline prow/plank/reconciler.go:891]: This would conflict with a user defined job queue with this name. We should just be able to use the empty string here, I don't think we need a magic value. If we do need a magic value for some reason, we should prevent users from using it themselves via config validation.

**Review by cjwagner — COMMENTED** [review]
Feel free to add tests and squash commits. Generally looks good to me besides missing the config validation.

File: prow/config/jobs.go
Comment (jupblb) [reply prow/config/jobs.go]: Changed.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: Removed.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: Done. Made `validateJobBase` a method of `Config`. Thank you for explaining the existing logic to me.

**Comment by jupblb** [comment]
Added the config validation and fixed other issues. I think this PR is ready for a (hopefully) final review round. :)

File: prow/plank/reconciler.go
Comment (stevekuznetsov) [inline prow/plank/reconciler.go:821]: Should this be `Debugf()`?

File: prow/plank/reconciler.go
Comment (stevekuznetsov) [inline prow/plank/reconciler.go:992]: Why do we know this about the state? Why is it not errored? Why not check explicitly?

File: prow/plank/reconciler.go
Comment (stevekuznetsov) [inline prow/plank/reconciler.go:994]: Do we allow UPDATE/PATCH to the spec to change the queue name?

File: prow/plank/reconciler_test.go
Comment (stevekuznetsov) [inline prow/plank/reconciler_test.go:316]: nit: we should not use this method in tests - provide the literal here so someone who changes the semantics of the index knows they've done so

File: prow/plank/reconciler.go
Comment (stevekuznetsov) [inline prow/plank/reconciler.go:995]: Do we want to break ties on timestamp ordering with lexicographic sort of UID or something?

**Review by stevekuznetsov — COMMENTED** [review]
Cool feature! Implementation looks pretty good, just a couple of questions.

File: prow/plank/reconciler_test.go
Comment (alvaroaleman) [reply prow/plank/reconciler_test.go:316]: This is actually fine, because the index only exists in Plank in memory, so if someone were to change its name or such, it would still work (as long as the correct name is referenced everywhere)

Comment (alvaroaleman) [reply prow/plank/reconciler_test.go:316]: Unless this is serialized somehow/needs to be forwards/backwards compatible using the same func to generate the key throughout production code and tests is best imho, because if it needs to be changed it doesn't require the tests to be updated. But it is not a big deal one way or the other

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go:821]: This code mimics the behaviour of `canExecuteConcurrentlyPerJob` (which was the original implementation): https://github.com/kubernetes/test-infra/blob/master/prow/plank/reconciler.go#L782. But I agree `Debugf` could work here just fine.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go:992]: Good catch. This is my bad. This function used to be called in context of using `optPendingTriggeredJobsNamed`. That is why there is no explicit check for the `Triggered` state.

I added the check and renamed function to `countPendingOrOlderTriggeredMatchingPJs`.

File: prow/plank/reconciler.go
Comment (jupblb) [reply prow/plank/reconciler.go]: It does not matter to me and it was not the case for the `MaxConcurrency` mechanism.

File: prow/plank/reconciler_test.go
Comment (jupblb) [reply prow/plank/reconciler_test.go:316]: I side with @alvaroaleman on this one, I think it's more practical to have this function called instead of having a separate literal. But if we wanted to have this literal after all I'd say this should be extracted to a separate task (to change it for the entire test suite) and is therefore outside of scope for this PR.

File: prow/plank/reconciler_test.go
Comment (alvaroaleman) [inline prow/plank/reconciler_test.go:362]: Can we please also add a testcase in `TestMaxConcurency` (in the prow/plank/controller_test.go file) to actually test the functionality rather than just testing the indexer change?

File: prow/plank/reconciler.go
Comment (alvaroaleman) [inline prow/plank/reconciler.go:893]: Maybe return an empty string here if `jobQeueName` is an empty string to avoid indexing jobs by queue that do not have a queue set?

File: prow/plank/reconciler_test.go
Comment (jupblb) [reply prow/plank/reconciler_test.go:362]: Good idea, implemented. Since `canExecuteConcurrentlyPerJob` and `canExecuteConcurrentlyPerQueue` share `countPendingOrOlderTriggeredMatchingPJs` logic I added only two tests:
- Verify that limit `0` means that there is no limit
- Verify that a job can't be executed if the job queue is full

**Comment by jupblb** [comment]
Thanks for the review. Fixed the new issues. The diff should(?) be accessible here: https://github.com/jupblb/test-infra/commit/e81db6c895ec90cce67fecb9aa619786b53520ec. :)

**Review by alvaroaleman — APPROVED** [review]
thank you!
/hold cancel

**Pull request merged**
